### ISSUE TITLE
Types on kwargs can lead to confusing situations

### ISSUE THREAD

**Issue by IainNZ** [JuliaLang/julia#16490]
I just helped @evanfields boil down a more complex situation to the following:

```
function trigger(a, b; c::Float64 = a * b)
    return true
end
```

On Julia 0.4, with `trigger(5, 10, c = 3)` you get
```
ERROR: LoadError: TypeError: trigger: in typeassert, expected Float64, got Int64
```
Which I consider fairly understandable (unfortunate that you don't get told which field it was though).

With `trigger(5, 10)`, again on 0.4, however, you get
```
ERROR: LoadError: MethodError: `__trigger#0__` has no method matching __trigger#0__(::Int64, ::Int64, ::Int64)
Closest candidates are:
  __trigger#0__(::Float64, ::Any, ::Any)
```
Which, while fairly understandable in retrospect, is totally baffling unless you know how `kwargs` are lowered. Its unfortunate that its a different error message though, for a similar sin. It is also hard to figure out whats going on in more complex cases - here it was a 2+1 arg function, but the original setting was much more interesting.

On Julia 0.5 you get
```
ERROR: LoadError: TypeError: #trigger: in typeassert, expected Float64, got Int64
 in (::#kw##trigger)(::Array{Any,1}, ::#trigger, ::Int64, ::Int64) at ./null:0
```
for the first, which is worse than 0.4 (the null:0 is delightful) , and for the second case you get
```
ERROR: LoadError: MethodError: no method matching #trigger#1(::Int64, ::#trigger, ::Int64, ::Int64)
Closest candidates are:
  #trigger#1(::Float64, ::Any, ::Any, ::Any)
 in trigger(::Int64, ::Int64) at /Users/idunning/....
```
which is even more exciting maybe.

Is there a path forward for making these messages nicer, or is too much information lost when lowering?

**Comment by IainNZ** [comment]
Related whining about error messages and kwargs in https://github.com/JuliaLang/julia/issues/15639

**Comment by andyferris** [comment]
I agree all of this is very confusing, and has caught me in the past. +1 for fixing - assuming it isn't too hard (sorry I wouldn't know where to start).

**Comment by IainNZ** [comment]
Still a problem IMO (behaviour still same as in 0.5 I think), but very much post 1.0.

**Comment by ExpandingMan** [comment]
I just got burned by this, minimal example on Julia 1.1.1 as follows:
```julia
julia> const A = nothing

julia> f(;a::Int=A) = a
f (generic function with 1 method)

julia> f()
ERROR: MethodError: no method matching #f#7(::Nothing, ::typeof(f))
Closest candidates are:
  #f#7(::Int64, ::Any) at REPL[2]:1
Stacktrace:
 [1] f() at ./REPL[2]:1
 [2] top-level scope at none:0
```
Took me quite a while to figure out what was going on when I encountered this in real code.
